map-guessr/index.ts

```diff
@@ -0,0 +1,813 @@
+  const regex = /(?<sign>[+-]?)\s*(?<number>\d+\.?\d*|\.\d+)/i;
+  const lat_result = str.match(lat_regex)
+    ? str.match(lat_regex).groups
+    : str.match(regex)
+    ? str.match(regex).groups
+    : (undefined as { sign: string; number: string } | undefined);
+  str = str.replace(
+    new RegExp(lat_result.sign + "\\s*" + lat_result.number, "i"),
+    ""
+  );
+  console.log(str);
+  const lng_result = str.match(lng_regex)
+    ? str.match(lng_regex).groups
+    : str.match(regex)
+    ? str.match(regex).groups
+    : (undefined as { sign: string; number: string } | undefined);
+  if (!lat_result) return null;
+  if (!lng_result) return null;
+  let lat = parseFloat(lat_result.number);
+  let lng = parseFloat(lng_result.number);
+  if (lat > 90 || lng > 180) return null;
+  const lat_sign = lat_result.sign.toLowerCase();
+  const lng_sign = lng_result.sign.toLowerCase();
+  if (lat_sign === "s" || lat_sign === "-") lat = -lat;
+  if (lng_sign === "w" || lng_sign === "-") lng = -lng;
+  return [lat, lng];
+}
+
+function latLngToMercator(lat: number, lng: number): number[] {
+  const x = (lng / 180) * Math.PI;
+  const y =
+    Math.log(
+      (1 + Math.sin((lat / 180) * Math.PI)) /
+        (1 - Math.sin((lat / 180) * Math.PI))
+    ) / 2;
+  return [x, y];
+}
+
+function polarToCartesian(phi: number, theta: number): number[] {
+  const x = Math.cos(phi) * Math.cos(theta);
+  const y = Math.cos(phi) * Math.sin(theta);
+  const z = Math.sin(phi);
+  return [x, y, z];
+}
+
+function randomPoint(size: number): number[] {
+  while (true) {
+    const x = Math.random() * 2 - 1;
+    const y = Math.random() * 2 - 1;
+    const z = Math.random() * 2 - 1;
+    const r = (x ** 2 + y ** 2 + z ** 2) ** 0.5;
+    if (0 < r && r < 1) {
+      const lat = (Math.asin(z / r) / Math.PI) * 180;
+      const lng = (Math.atan2(x / r, y / r) / Math.PI) * 180;
+      if (Math.abs(lat) < 80 - size / 2 / ((6378.137 * Math.PI) / 180)) {
+        return [lat, lng];
+      }
+    }
+  }
+}
+
+function distFormat(num: number): string {
+  if (num >= 10) {
+    return `${Math.round(num)}km`;
+  } else if (num >= 1) {
+    return `${num.toFixed(1)}km`;
+  } else {
+    return `${Math.round(num * 1000)}m`;
+  }
+}
+
+export default async ({ eventClient, webClient: slack }: SlackInterface) => {
+  const aliases = (await fs.readJson(
+    __dirname + "/country_names.json"
+  )) as Record<string, string[]>;
+
+  const countriesList = countriesExpand(["世界"], aliases)
+    .filter((country) => country !== "-")
+    .sort();
+
+  const aliasesStringArray = countriesListGen(aliases);
+
+  eventClient.on("message", async (message) => {
+    if (
+      message.channel !== CHANNEL ||
+      message.thread_ts ||
+      !(
+        message.text?.startsWith("座標当て") ||
+        message.text?.startsWith("座標あて")
+      )
+    ) {
+      return;
+    }
+
+    const messageTs = JSON.parse(JSON.stringify({ thread_ts: message.ts }));
```

Comment:
[MAY] objectのクローンをしているけど、かなり不要そう

```suggestion
    const messageTs = { thread_ts: message.ts };
```